<a href="https://colab.research.google.com/github/KaranSingh36752/DeepFake_Detection/blob/main/DistilBertFinetune.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install transformers datasets pandas

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 487.4/487.4 kB 11.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 7.7 MB/s eta 0:00:00


install Hugging Face's tools (transformers for models datasets for datasets) and Pandas for data handling.

In [ ]:
!pip install --upgrade torch


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 32.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 19.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 18.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 2.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 25.1 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling nvidia-nvjitlink-cu12-12.5.82:
      Successfully uninstalled nvidia-nvjitlink

In [ ]:
from datasets import load_dataset
dataset = load_dataset("go_emotions")
print("Dataset sample:", dataset["train"][0])
label_names = dataset["train"].features["labels"].feature.names
print(label_names)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/9.40k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/2.77M [00:00<?, ?B/s]

validation-00000-of-00001.parquet:   0%|          | 0.00/350k [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/347k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/43410 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/5426 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/5427 [00:00<?, ? examples/s]

Dataset sample: {'text': "My favourite food is anything I didn't have to cook myself.", 'labels': [27], 'id': 'eebbqej'}
['admiration', 'amusement', 'anger', 'annoyance', 'approval', 'caring', 'confusion', 'curiosity', 'desire', 'disappointment', 'disapproval', 'disgust', 'embarrassment', 'excitement', 'fear', 'gratitude', 'grief', 'joy', 'love', 'nervousness', 'optimism', 'pride', 'realization', 'relief', 'remorse', 'sadness', 'surprise', 'neutral']


Fetches a dataset of Reddit comments labeled with emotions like "joy," "anger," etc.


In [ ]:
import pandas as pd
df = pd.DataFrame(dataset["train"])
mood_map = {
    0: "happy",   # admiration → happy
    8: "happy",   # desire → happy
    23: "happy",  # relief → happy

    # Keep other mappings
    1: "happy",   # amusement
    4: "happy",   # approval
    5: "happy",   # caring
    13: "happy",  # excitement
    15: "happy",  # gratitude
    17: "happy",  # joy
    18: "happy",  # love
    20: "happy",  # optimism
    21: "happy",  # pride

    9: "sad",     # disappointment
    16: "sad",    # grief
    24: "sad",    # remorse
    25: "sad",    # sadness

    2: "angry",   # anger
    3: "angry",   # annoyance
    10: "angry",  # disapproval
    11: "angry",  # disgust

    6: "fearful",  # confusion
    12: "fearful", # embarrassment
    14: "fearful", # fear
    19: "fearful", # nervousness

    22: "surprised",  # realization
    26: "surprised",  # surprise

    # Reduced neutral mappings
    7: "neutral",   # curiosity
    27: "neutral"   # explicit neutral
}

def map_mood(labels):
    """Map labels to a single mood, defaulting to 'neutral' if no match."""
    for label in labels:
        if label in mood_map:
            return mood_map[label]
    return "neutral"

df["mood"] = df["labels"].apply(map_mood)

df = df[["text", "mood"]].sample(n=10000, random_state=42)
print("Mood distribution:\n", df["mood"].value_counts())

Mood distribution:
 mood
happy        3897
neutral      3393
angry        1346
sad           550
fearful       482
surprised     332
Name: count, dtype: int64


Groups complex emotions into 6 basic moods. Example: "joy" becomes "happy"

In [ ]:
from transformers import DistilBertTokenizer
tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
text = "I’m feeling great!"
tokens = tokenizer(text, padding="max_length", truncation=True,max_length=128,return_tensors="pt")
print(tokens)

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

{'input_ids': tensor([[ 101, 1045, 1521, 1049, 3110, 2307,  999,  102,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0,

Converts sentences into tokens that DistilBERT understands.
Example: "great" → [2307].

In [ ]:
import torch
from torch import nn
from transformers import DistilBertForSequenceClassification, Trainer, TrainingArguments

MOOD_LABELS = ["happy", "sad", "angry", "fearful", "surprised", "neutral"]

model = DistilBertForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=len(MOOD_LABELS), #moods
    id2label={i: label for i, label in enumerate(MOOD_LABELS)},
     label2id={label: i for i, label in enumerate(MOOD_LABELS)}
)

for param in model.distilbert.parameters():
    param.requires_grad = True



train_df = df.sample(frac=0.8, random_state=42)
val_df = df.drop(train_df.index)


train_encodings = tokenizer(train_df["text"].tolist(), truncation=True, padding=True, max_length=128)
val_encodings = tokenizer(val_df["text"].tolist(), truncation=True, padding=True, max_length=128)

train_labels = [model.config.label2id[mood] for mood in train_df["mood"]]
val_labels = [model.config.label2id[mood] for mood in val_df["mood"]]

class MoodDataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item["labels"] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)

train_dataset = MoodDataset(train_encodings, train_labels)
val_dataset = MoodDataset(val_encodings, val_labels)

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
from torch import nn
from transformers import Trainer

class_weights = torch.tensor(
    [1.0,  # happy
     10.0,  # sad
     5.0,  # angry
     8.0,  # fearful
     12.0,  # surprised
     0.5   # neutral
], dtype=torch.float32).to("cpu")

class CustomTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False,**kwargs):
        labels = inputs.pop("labels")
        outputs = model(**inputs)
        loss_fct = nn.CrossEntropyLoss(weight=class_weights)
        loss = loss_fct(outputs.logits, labels)
        return (loss, outputs) if return_outputs else loss

In [ ]:
from sklearn.metrics import classification_report

def compute_metrics(pred):
    labels = pred.label_ids
    preds = pred.predictions.argmax(-1)
    return classification_report(labels, preds, target_names=MOOD_LABELS, output_dict=True)

In [ ]:
training_args = TrainingArguments(
    output_dir="./results",
    evaluation_strategy="epoch",
    save_strategy="epoch",
    learning_rate=5e-5,  # Increased from 3e-5
    per_device_train_batch_size=32,
    num_train_epochs=5,
    weight_decay=0.01,
    fp16=False,
    gradient_accumulation_steps=1,
    warmup_ratio=0.2,
    gradient_checkpointing=True,
    load_best_model_at_end=True,
    metric_for_best_model="eval_loss",
    greater_is_better=False,
    logging_steps=50,
    warmup_steps=100,
    lr_scheduler_type="cosine",
)


trainer = CustomTrainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=compute_metrics
)

# Add EarlyStoppingCallback
from transformers import EarlyStoppingCallback
trainer.add_callback(EarlyStoppingCallback(
    early_stopping_patience=3,  # Stop if no improvement for 2 epochs
    early_stopping_threshold=0.01
))

trainer.train()

/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.


<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter:

 ··········


wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: karansingh36752 (karansingh36752-karansingh) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


KeyboardInterrupt: 

DistilBertForSequenceClassification: Loads DistilBERT and adds a classification layer for moods.

Trainer: Handles the training loop (forward pass, loss calculation, backpropagation).

TrainingArguments: Sets hyperparameters like learning rate and batch size.

In [ ]:
model.save_pretrained("./mood_model")
tokenizer.save_pretrained("./mood_model")

In [ ]:
!zip -r mood_model.zip mood_model/

In [ ]:
from google.colab import files
files.download('mood_model.zip')